presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.reservoirsample;
+
+import java.util.Arrays;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.DoubleStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class UnweightedDoubleReservoirSample
+        implements Cloneable
+{
+    protected int maxSamples;
+    protected long count;
+    double[] elements;
+
+    public UnweightedDoubleReservoirSample(int maxSamples)
+    {
+        this.maxSamples = maxSamples;
+        checkArguments();
+        this.elements = new double[0];
+    }
+
+    protected UnweightedDoubleReservoirSample(UnweightedDoubleReservoirSample other)
+    {
+        this.maxSamples = other.maxSamples;
+        checkArguments();
+        this.count = other.count;
+        this.elements = new double[other.elements.length];
+        System.arraycopy(other.elements, 0, elements, 0, elements.length);
+    }
+
+    public long getMaxSamples()
+    {
+        return maxSamples;
+    }
+
+    public void add(double element)
+    {
+        int index = getInsertIndex();
+        if (index >= 0) {
+            putAtIndex(index, Double.valueOf(element));
+        }
+    }
+
+    protected int getInsertIndex()
+    {
+        ++count;
+        if (elements.length < maxSamples) {
+            return elements.length;
+        }
+
+        int index = (int) ThreadLocalRandom.current().nextInt(0, (int) count);
+        System.out.println(index + " " + count);
+        return index < elements.length ? index : -1;
+    }
+
+    protected void putAtIndex(int index, double element)
+    {
+        if (index == -1) {
+            return;
+        }
+
+        if (elements.length < maxSamples) {
+            double[] newElements = new double[elements.length + 1];
```

Comment:
- Re-allocating the array and copying data for each element is wasteful; why not allocate maxSamples in the constructor?

```
    public UnweightedDoubleReservoirSample(int maxSamples)
    {
        checkArgument(maxSamples > 0, "Number of reservoir samples must be strictly positive");
        checkArgument(maxSamples < MAX_SAMPLES_LIMIT, "Number of reservoir samples must not exceed " + MAX_SAMPLES_LIMIT);
        this.maxSamples = maxSamples;
        this.elements = new double[maxSamples];
    }
```